pkg/apis/mongodb/v1/mongodb_types.go

```diff
@@ -42,6 +45,29 @@ type MongoDBSpec struct {
 	// be set for the deployment
 	// +optional
 	FeatureCompatibilityVersion string `json:"featureCompatibilityVersion,omitempty"`
+
+	// TLS configuration for both client-server and server-server communication
+	// +optional
+	TLS TLS `json:"tls"`
+}
+
+// TLS is the configuration used to set up TLS encryption
+type TLS struct {
+	Enabled bool `json:"enabled"`
+
+	// Optional configures if TLS should be required or optional for connections
+	// +optional
+	Optional bool `json:"optional"`
```

Comment:
How would a user handle the case of upgrading an existing cluster to TLS and upgrading their clients to TLS as well? When TLS is enabled and becomes required, existing non-TLS clients would get disconnected. The documentation recommends doing this as part of the upgrade procedure in step 2: https://docs.mongodb.com/manual/tutorial/upgrade-cluster-to-ssl/.

Reply:
I think you are right, although I think it is unfortunate the way we phrase this. Docs should be "make sure all of your clients can work with both TLS and non-TLS enabled clusters and then upgrade the cluster to TLS only.

This is exactly the same as having an HTTPS server that also handles HTTP in case the client does not support HTTPS. which I don't think anyone would agree that it is a good idea.

Anyway, I think I'm fine with having optional in here

Reply:
The recommendation might stem from the fact that a lot of drivers don't seem to support automatic switching between TLS and non-TLS. Nevertheless, we are probably best of following the official recommendations.